src/tfr_utilities/src/obstacle_broadcaster.cpp

```diff
@@ -0,0 +1,108 @@
+#include <obstacle_broadcaster.h>
+
+/*
+ * Initializes the broadcaster and data structures
+ * */
+ObstacleBroadcaster::ObstacleBroadcaster(
+        ros::NodeHandle& n, 
+        const std::string &point_frame, 
+        const std::string &parent_frame, 
+        const std::string &service,
+        const double &d) : 
+
+    node{n}, 
+    broadcaster_frame{point_frame},
+    map_frame{parent_frame}, 
+    service_name{service},
+    diameter{d},
+    point_set{false}
+{
+    server = node.advertiseService(service_name,
+            &ObstacleBroadcaster::localizePoint, this);
+    transform.header.frame_id = map_frame;
+    transform.child_frame_id = broadcaster_frame;
+    transform.transform.rotation.w = 1;
+    /* NOTE we work with the pcl::PointCloud class and broadcast it as a
+     * sensor_msgs::PointCloud2, this is intentional, and in fact the way it is
+     * designed to be used in ROS
+     * */
+    cloud_publisher = node.advertise<sensor_msgs::PointCloud2>("obstacle_cloud", 5);
+}
+
+/*
+ * Broadcasts point across the transform network
+ * If the point has been set it will also broadcast the obstacle
+ * */
+void ObstacleBroadcaster::broadcast()
+{
+    transform.header.stamp = ros::Time::now();
+    broadcaster.sendTransform(transform);
+    if (point_set)
+    {
+        pcl::PointCloud<pcl::PointXYZ> cloud;
+        generateCloud(cloud);
+        cloud_publisher.publish(cloud);
+    }
+}
+
+/*
+ * Gives the point a new origin
+ * */
+bool ObstacleBroadcaster::localizePoint(tfr_msgs::LocalizePoint::Request &request,
+        tfr_msgs::LocalizePoint::Response &resonse)
+{
+    transform.transform.translation.x = request.pose.pose.position.x;
+    transform.transform.translation.y = request.pose.pose.position.y;
+    transform.transform.translation.z = request.pose.pose.position.z;
+    transform.transform.rotation.x = request.pose.pose.orientation.x;
+    transform.transform.rotation.y = request.pose.pose.orientation.y;
+    transform.transform.rotation.z = request.pose.pose.orientation.z;
+    transform.transform.rotation.w = request.pose.pose.orientation.w;
+    point_set = true;
+    return true;
+}
+
+/*
+ *  Generates a circular pointcloud levitating above the ground around the hole,
+ *  used to make sure the robot, never ever falls in.
+ * */
+void ObstacleBroadcaster::generateCloud(pcl::PointCloud<pcl::PointXYZ> &cloud)
+{
+    double PI = 3.14159;
+    cloud.header.frame_id = broadcaster_frame;
+    cloud.width    = 16;
+    cloud.height   = 1;
+    cloud.is_dense = true;
+    double radius = diameter/2, offset = 0.0;
+    for (int i = 0;  i < 16; i++, offset += PI/8) 
+        cloud.points.push_back(pcl::PointXYZ( radius*cos(offset), radius*sin(offset), 0.15));
+
```

Comment:
I'd like you to define a constant for the width, move it to the top of the file, and add a TODO comment to tweak it once we know for sure what size of hole we expect to leave behind.